fixes #26741 - adds containers support for developers

ohadlevy opened the pull request:
This patch introduce a docker file and docker-compose example so that developers can easily get a 
production docker based environment working quickly.

This by no means is a replacement to the installer or should effect users using Foreman besides the propose of getting up an environment quickly.

This does not try to solve docker problems for plugins (but it can be used to build customized images with a custom plugin installed).

To test this pr you would need docker and docker-compose (I'm running docker-compose-1.22.0-2.)

to use this PR:
first start by polling the images
```
docker-compose pull
```
followed by a db create/migrate etc (as its a new env)
side note - I've used MySQL as that's the environment I use in production, I can share how to use PG instead of even change it to PG by default if desired.

```
docker-compose run app bundle exec rake db:create db:migrate
docker-compose run app bundle exec rake db:seed
```

finally start it
```
docker-compose up
```
You should know be able to login to ```http://localhost:3000```

additionally, this image contains a report of the webpack build that can be found at  ```http://localhost:3000/webpack/report.html```

This will run 3 different containers, one for the DB, another for foreman and the last one for dynflow.

In order to build your own docker image you can either use a service (such as quay.io - the image used in this PR has been built at https://quay.io/repository/ohadlevy/foreman/build/676ac307-a838-4e91-988b-76be2ddb7680) triggered by a git push or manually OR install a recent version of docker that supports multi-stage builds and execute

```
docker-compose build
```
in theory podman also supports it, but i had stability issues with it while trying to build images.

ohadlevy committed:
fixes #26741 - adds containers support for developers

this patch introduce a docker file and docker compose example
so that developers can easily get a production docker based env
working quickly.

this by no means is a replacement to the installer or should
effect users besides the porpose of getting up an environment quickly

this does not try to solve docker problems for plugins (but it
can be used to build customized images with a custom plugin installed).

to test this pr you would need docker and docker-compose (I'm running docker-compose-1.22.0-2.)

to use this PR:
(install docker-compose)

docker-compose run app bundle exec rake db:create db:migrate db:seed
docker-compose up

in order to build your own docker image you can either use a service
(such as quay.io) tiggered by a git push or manually OR
to install a recent version of docker that supports multi-stage builds

in theory podman also supports it, but i had stability issues with it
while trying to build images.

timogoebel commented on .dockerignore:
```diff
@@ -0,0 +1,57 @@
+.gitignore
+.dockerignore
+.travis.yml
+kube
```
??

ohadlevy replied:
I'm sorry, that's a leftover from from another work (if you are interested you can see it at https://github.com/ohadlevy/foreman/tree/dockerfile/kube) will remove..

timogoebel commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+# Base container that is used for both building and running the app
+FROM centos:latest as base
+ENV RUBY_SCL="rh-ruby25"
```
The shipped ruby version is quite outdated and has known bugs. There is a bugzilla to upgrade to the latest ruby version: https://bugzilla.redhat.com/show_bug.cgi?id=1688754
Can we use a base image that ships a newer ruby version?

ohadlevy replied:
while that's possible, I was trying to reduce deviation form the current ruby versions we use (for test matrix etc) regardless, I didn't use a ruby based image as we also need nodejs.

ohadlevy replied:
I went ahead and used fedora:30 with ruby 2.6 / npm 11 using dnf modules, i left it outside of this PR at https://github.com/ohadlevy/foreman/commit/d0e030854565851c6d280a968550aecc282240b7 because i didn't use it yet (just confirmed the app loads correctly).

if this is the direction you had in mind let me know and I'll continue working on that branch, the bonus points is that it will be compatible with centos 8 once its released..

timogoebel commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+    chown -R 1001:1001 $HOME
+
+# Add a script to be executed every time the container starts.
+COPY entrypoint.sh /usr/bin/
```
We could get rid of this if we don't use the SCL. That's a huge gain IMHO because it simplifies the usage.

ohadlevy replied:
TBH that's not just for SCL, i think its useful to ensure a certain script runs before other commands...

timogoebel commented on docker-compose.yml:
```diff
@@ -0,0 +1,55 @@
+      - db:/var/lib/mysql/data
+
+  app: &app_base
+    image: quay.io/ohadlevy/foreman:containers
```
Does this need to be adjusted?

ohadlevy replied:
I'm happy to change it if poeple agree, I wanted to have a working container people can try out this pr...

ohadlevy replied:
I'll adjust it once merged?

timogoebel replied:
You mean in a follow-up commit. 👍 from me.

timogoebel reviewed (changes requested):
I think this is great to have and we should definitely start shipping Docker containers for the App. We can iterate on this and add more features later. But having a container just for a vanilla foreman is a huge gain. If users need plugins, they can build their own container off of this for now.

timogoebel commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+ENV FOREMAN_APIPIE_LANGS="en"
+
+ENV HOME=/home/foreman
+ARG BUNDLER_SKIPPED_GROUPS="mysql test development jenkins openid libvirt journald"
```
Why do you skip `mysql` here when your database is in fact MySQL?

ohadlevy replied:
its the mysql (1) gem, everyone moved to mysql2 :) I guess we can remove it in core as I dont really assume it still works?

ekohl replied:
The `mysql` group was removed in 63b3ed2d1d5e95af8fedb3fab8cdde3357a0eafd (dated in 2013). There is also no jenkins group.

timogoebel commented:
@ohadlevy: Are you aware of https://github.com/theforeman/foreman/pull/5207? That's a more lightweight solution which I honestly prefer.

ohadlevy commented:
> @ohadlevy: Are you aware of #5207? That's a more lightweight solution which I honestly prefer.

I was not! sadly i didn't see it before starting. and I think it actually completes this PR, as this work is intended towards production builds and AFAIU #5207 is aimed at development environment ? i could see a second docker compose file added to the repo to support those usage cases too?

ohadlevy commented:
@timogoebel or others any other thoughts? I would be happy to promote this further.

timogoebel commented:
Please do promote this further. I‘ll be happy to review and merge. We don‘t have to solve every use case in the first iteration. But having a base container image would be great.

ohadlevy commented:
Just for me to understand what's the next step, when you say base image you
mean a dev environment? Something else? TBH for me it's already useful as
is.

On Sat, May 4, 2019, 12:03 Timo Goebel <notifications@github.com> wrote:

> Please do promote this further. I‘ll be happy to review and merge. We
> don‘t have to solve every use case in the first iteration. But having a
> base container image would be great.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/6730#issuecomment-489308895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAE5R5E5J3H3G37RJVDXXDPTVGOVANCNFSM4HJ4SB7A>
> .
>

timogoebel commented:
> it's already useful as is.

That's what I'm trying to say. Let's merge this and iterate on it.

> when you say base image you mean a dev environment

I mean base image as in something we can start iterating on. :-)

ohadlevy commented:
> That's what I'm trying to say. Let's merge this and iterate on it.

OK, so I've updated to use latest fedora with the aim to swap to centos 8 once its released.

I would also like to enable quay.io image building for foreman branches (develop / stable) once we have this PR merged.

ohadlevy commented:
build of this image log can be seen at https://quay.io/repository/ohadlevy/foreman/build/721c2805-473d-4f36-805c-33a947ce4449

timogoebel commented on docker-compose.yml:
```diff
@@ -0,0 +1,55 @@
+    environment:
+      - DATABASE_URL=mysql2://foreman:foreman@db:3306/foreman
+      - RAILS_MAX_THREADS=5
+      #- SECRET_KEY_BASE=my-secret-access-key
```
Do we need this?

timogoebel commented on docker-compose.yml:
```diff
@@ -0,0 +1,55 @@
+  worker:
+    <<: *app_base
+    command: bundle exec rake dynflow:executor
+    hostname: worker.docker-rails.my-site.com
```
Same here: `example.com`?

ekohl commented on Dockerfile:
```diff
@@ -0,0 +1,95 @@
+RUN mkdir bundler.d && mkdir config
+COPY Gemfile ${HOME}
+COPY bundler.d/* bundler.d/
+RUN echo gem \"rdoc\" > bundler.d/container.rb
```
I also need to add this on my Fedora. Do we know why? Should this just be part of the regular development setup?

ohadlevy replied:
I assume that it doesn't get required as its not in the bundle file, need to see how apipie uses it.. it was a quick workaround when I moved to fedora (or recent rubies?) based images.

ekohl commented on Dockerfile:
```diff
@@ -0,0 +1,95 @@
+  find vendor/ruby/*/gems -name "*.o" -delete
+COPY package.json ${HOME}
+RUN entrypoint.sh npm install --ignore-scripts --no-optional
+COPY --chown=1001:0 . ${HOME}/
```
You also added a group foreman so why not `--chown=1001:1001`?

ohadlevy replied:
I was thinking of using it in openshift as well, where the user that execute the image is random, I'll change the group to foreman too.

ekohl commented on docker-compose.yml:
```diff
@@ -0,0 +1,55 @@
+      - MYSQL_PASSWORD=foreman
+      - MYSQL_DATABASE=foreman
+      - MYSQL_DATADIR_ACTION=upgrade-auto
+    hostname: db.prod.com
```
I'd use `example.com`.

ohadlevy commented:
updated, image built log at https://quay.io/repository/ohadlevy/foreman/build/6d7dc1f5-6d56-4089-8b3d-e90ef139c16d

boaz0 commented on Dockerfile:
```diff
@@ -0,0 +1,95 @@
+  dnf clean all && \
+  rm -rf /var/cache/dnf/
+
+ENV RAILS_ENV="production"
```
I see that you use this ENV for both build-stages. Did you consider this: https://github.com/moby/moby/issues/37345#issuecomment-400245466?

All in all, it looks good to me and I believe the image size was reduced drastically.

ohadlevy replied:
@boaz0 thanks, I did try that, but it seems the docker version that we have on quay.io doesnt support that (https://quay.io/repository/ohadlevy/foreman/build/395a5fa8-f543-4cae-8731-54a24acd9730).

I did do minor cleanups on the ARGS but they are still duplicated across the images.

timogoebel commented:
What I found while testing:

* `docker-compose up` has to be run before migrate & seed
* `rake db:create db:migrate` and `db:seed` need to be separate

ohadlevy commented:
> What I found while testing:
> 
> * `docker-compose up` has to be run before migrate & seed

does docker-compose run instead of exec work as expected?
> * `rake db:create db:migrate` and `db:seed` need to be separate

this is a known issue (or so @tbrisker told me, I'm not sure which issue it should be linked to)

timogoebel commented on docker-compose.yml:
```diff
@@ -0,0 +1,54 @@
+
+  app: &app_base
+    image: quay.io/ohadlevy/foreman:containers
+    command: bundle exec bin/rails server
```
I had to change this to `command: bundle exec bin/rails server -b 0.0.0.0` to access the app.

timogoebel commented:
I'm currently stuck at `Webpack::Rails::Manifest::ManifestLoadError`. Will investigate.

ehelms commented:
Weighing in so these points are at least considered:

 * There are no tests for this, and experience has taught me that leads to unstable environments
 * Packaging living in code repositories has historically proven to be painful in time for large projects
 * This will likely need a packaging update to exclude these new files

timogoebel commented:
@ehelms:

> * There are no tests for this, and experience has taught me that leads to unstable environments

What kind of tests do you want to have? That the container builds will be verified by quay.io.
Integration tests of the app in the container would be nice to have, but I believe we can add them later.

> * Packaging living in code repositories has historically proven to be painful in time for large projects

Packaging living in a separate repo has been proven to be painful for this particular project as well. Having the code in this repo increases visibility and we don't have to synchronize multiple repos. In addition, it's clear that it's the core maintainers job to keep the container working.
We can easily change this at a later point if we decide that it was a bad idea.

> * This will likely need a packaging update to exclude these new files

That's good to know. Who do you think should take care of making the necessary changes?

ohadlevy commented:
@timogoebel updated
* bind address to include 0.0.0.0 by default
* fixed permissions to be consistent with all copies (e.g. I've noticed some files were owned by root)
* added python (was missing in fedora 30 image) to allow the vnc websockets script to run
* added gettext:po_to_json rake task so JS translation strings are available.

ohadlevy commented:
status update
* rebased
* fixed JS gettext by using the right task
* remove node_modules from plugins as well (vs just foreman).

IMHO ready to merge :-)

boaz0 commented:
LGTM :+1:

ehelms commented:
@timogoebel The situation I envision is. A code change is made. That code change breaks the container build. A developer goes to pull the latest container and now has a container that doesn't work with current code because the build broke and a new container was not created. A developer now has to track down why things are broken. This is the cycle our current devel environments run into.

ohadlevy commented:
> @timogoebel The situation I envision is. A code change is made. That code change breaks the container build. A developer goes to pull the latest container and now has a container that doesn't work with current code because the build broke and a new container was not created. A developer now has to track down why things are broken. This is the cycle our current devel environments run into.

@ehelms I think your approach is valid, after all, my goal in this PR is to make developer life easier to reproduce production issues.

tl;dr Lets merge this and have follow up discussions on how to use this new tool.

I don't think this PR resolve every issue around containers, in fact, it barely scratch the surface and I think we should not force workflow discussions into it, or in other words, I think this enable a new discussion and should not block the PR, a few topics I could think of: (which probably should be in discourse)
* which base image should we use, maybe [UBI](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/building_running_and_managing_containers/index?lb_target=stage&extIdCarryOver=true&sc_cid=701f2000001Css5AAC#using_red_hat_universal_base_images_standard_minimal_and_runtimes) - something like [this](https://github.com/ohadlevy/foreman/tree/containers-ubi) or better use ubi-minimal
* should github/jenkins track quay build status in the PR level or only post merge? perhaps both?
* which kind of containers should we build by default (e.g. which plugin combinations are useful?)
* should we run E2E UI tests on top of the containers (it is fairly fast to build - maybe we could use it during PR vs after merge?)

overall, I've been building this container multiple times a day over the last month or two, there were no changes that happened in foreman that required a change to the container... I say lets merge it but not block any PR because of it, if its indeed useful as we all think it is, than we can resolve those issues just like we resolve any other issue....?

ekohl commented:
> which base image should we use, maybe [UBI](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/building_running_and_managing_containers/index?lb_target=stage&extIdCarryOver=true&sc_cid=701f2000001Css5AAC#using_red_hat_universal_base_images_standard_minimal_and_runtimes) - something like [this](https://github.com/ohadlevy/foreman/tree/containers-ubi) or better use ubi-minimal

+1 for UBI

> should github/jenkins track quay build status in the PR level or only post merge? perhaps both?

We have [foreman-develop-package-release](https://ci.theforeman.org/job/foreman-develop-package-release/) which releases to both RPM and deb. IMHO a third branch to build in Quay is the most logical place. It gives us a clear place to track.

> which kind of containers should we build by default (e.g. which plugin combinations are useful?)

Perhaps the various plugins could be a build parameter so users who wish to build with plugins can easily do so themselves but let's start with vanilla containers.

> should we run E2E UI tests on top of the containers (it is fairly fast to build - maybe we could use it during PR vs after merge?)

Do we have any E2E UI tests already? I'm not aware of any.

ohadlevy commented:
> > which base image should we use, maybe [UBI](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/building_running_and_managing_containers/index?lb_target=stage&extIdCarryOver=true&sc_cid=701f2000001Css5AAC#using_red_hat_universal_base_images_standard_minimal_and_runtimes) - something like [this](https://github.com/ohadlevy/foreman/tree/containers-ubi) or better use ubi-minimal
> 
> +1 for UBI
> 
do others feel like that? I can swap to the above branch, but it has older rubies and node(e.g. 2.5 and 10).

> > should github/jenkins track quay build status in the PR level or only post merge? perhaps both?
> 
> We have [foreman-develop-package-release](https://ci.theforeman.org/job/foreman-develop-package-release/) which releases to both RPM and deb. IMHO a third branch to build in Quay is the most logical place. It gives us a clear place to track.
> 

can we try this out before making it too complicated? it is intended for developers or people very familiar with containers, I don't think we should target main stream users.
Also, having it in the same repo and having docker-compose up, or kubectl create -f foreman.yaml can reduce friction.

Also, what if I'm wrong and this is not useful? why go though all of the release process and extra work at this point? lets proof its useful first?

I would suggest, that if at a later time we would like to restructure how we build the containers, we could move the dockerfile out side of the repo, but for now, IMHO there is more benefits of keeping it in.

> > which kind of containers should we build by default (e.g. which plugin combinations are useful?)
> 
> Perhaps the various plugins could be a build parameter so users who wish to build with plugins can easily do so themselves but let's start with vanilla containers.
> 
> > should we run E2E UI tests on top of the containers (it is fairly fast to build - maybe we could use it during PR vs after merge?)
> 
> Do we have any E2E UI tests already? I'm not aware of any.

I was referring to work [done](https://community.theforeman.org/t/rfc-integration-tests-end-to-end-tests/12142) by @amirfefer

ekohl commented:
> can we try this out before making it too complicated? it is intended for developers or people very familiar with containers, I don't think we should target main stream users.

We want to automate building of the container somewhere. Where else would you want to do this that also gives visibility to it?

Note that the job as an input has a git clone/tarball of verified sources. We've ran all the integration tests against them as regular CI jobs. Then for RPMs it basically does `obal release`. For Debs it does the equivalent. For the container step it could do a `docker build && docker push`. AFAIK that's all we need to do (modulo some credentials), right?

ohadlevy commented:
> We want to automate building of the container somewhere. Where else would you want to do this that also gives visibility to it?
> 
> Note that the job as an input has a git clone/tarball of verified sources. We've ran all the integration tests against them as regular CI jobs. Then for RPMs it basically does `obal release`. For Debs it does the equivalent. For the container step it could do a `docker build && docker push`. AFAIK that's all we need to do (modulo some credentials), right?

I was initially thinking of just using quay fast builders to run the builds for us, they are triggered by git merge, but I'm fine if we want to use another process. again, my aim in this PR is to get the ability to build a container, all the rest imho can follow up afterwards.

the main difference in thinking imho is that I want developers to be building containers (or using a service like quay/dockerhub etc) so that others can just pull their container ready to be tested, imho this can simplify the ability for non developers (users, ux etc) to simply get a working env quickly. hence I'm less thinking of how do we release it, vs give developers the ability to start using containers as something you can trade around.

ekohl commented:
> the main difference in thinking imho is that I want developers to be building containers (or using a service like quay/dockerhub etc) so that others can just pull their container ready to be tested, imho this can simplify the ability for non developers (users, ux etc) to simply get a working env quickly. hence I'm less thinking of how do we release it, vs give developers the ability to start using containers as something you can trade around.

Now that's something I kind of disagree with unless we decide to formally kill plugins (which the UI refactor already started doing) or find a way to support plugins. The reason is that it's one more step in the direction of not supporting plugins which will further split the ecosystem. As a release engineer, this is driving me insane and making it even easier for developer to not care about plugins is making my life hard. The past few releases we've always had delays because some UI feature didn't consider plugins and we needed to fix it. The argument that it's temporary or just an experiment is also weak because nothing is as permanent as something temporary and experiments always end up breaking other things. Merging it into core is giving the impression that it's production ready. Users will assume so because there's no indication whatsoever that this is an experiment.

ohadlevy commented:
> > the main difference in thinking imho is that I want developers to be building containers (or using a service like quay/dockerhub etc) so that others can just pull their container ready to be tested, imho this can simplify the ability for non developers (users, ux etc) to simply get a working env quickly. hence I'm less thinking of how do we release it, vs give developers the ability to start using containers as something you can trade around.
> 
> Now that's something I kind of disagree with unless we decide to formally kill plugins (which the UI refactor already started doing) or find a way to support plugins. The reason is that it's one more step in the direction of not supporting plugins which will further split the ecosystem. As a release engineer, this is driving me insane and making it even easier for developer to not care about plugins is making my life hard. The past few releases we've always had delays because some UI feature didn't consider plugins and we needed to fix it. The argument that it's temporary or just an experiment is also weak because nothing is as permanent as something temporary and experiments always end up breaking other things. Merging it into core is giving the impression that it's production ready. Users will assume so because there's no indication whatsoever that this is an experiment.

Why do think this works only for core? (excluding Katello that needs a complete different docker-compose file?) you can build easily any plugin configuration you like (from a local repo checkout that has plugin enabled).

ekohl commented:
> Why do think this works only for core? (excluding Katello that needs a complete different docker-compose file?) you can build easily any plugin configuration you like (from a local repo checkout that has plugin enabled).

Because I saw nothing mentioning plugins, but I now realize that by copying the bundler files you can. One common issue will be when people use relative paths to another git checkout. Any thoughts on handling that?

ohadlevy commented:
> One common issue will be when people use relative paths to another git checkout. Any thoughts on handling that?

You are correct, one workaround I can think of is to expose the git checkout directory so its available to the building container, but I've yet to experiment with that.

ShimShtein commented:
@ekohl 
> Because I saw nothing mentioning plugins, but I now realize that by copying the bundler files you can. One common issue will be when people use relative paths to another git checkout. Any thoughts on handling that?

``` sh
 tar cf - --exclude-from=.dockerignore . ../foreman_MY_PLUGIN  | docker build -
```

Don't forget to remove `Dockerfile` from `.dockerignore`. Assuming your plugins live in "sibling" folders to your core.

ohadlevy commented:
> @ekohl
> 
> > Because I saw nothing mentioning plugins, but I now realize that by copying the bundler files you can. One common issue will be when people use relative paths to another git checkout. Any thoughts on handling that?
> 
> ```shell
>  tar cf - --exclude-from=.dockerignore . ../foreman_MY_PLUGIN  | docker build -
> ```
> 
> Don't forget to remove `Dockerfile` from `.dockerignore`. Assuming your plugins live in "sibling" folders to your core.

@ekohl @ShimShtein 

if you don't want to mess up with paths, I would suggest to create directory called "dev" or something under your foreman directory, so in your bundle file you have something like
```rb
gem 'foreman_memcache', path: 'dev/foreman_memcache'
```
I personally just did
```sh
mkdir dev
ln -s ../../foreman_memcache .
cd ..
```

and then, let tar follow symlinks, e.g. 
```sh
tar -chf - --exclude-from=.dockerignore | docker build -
```

ohadlevy commented:
* rebased
* excluded tmp and coverage dirs
* added back dockerfile into the container so it can be use by the above tar workaround for path based gems.

timogoebel reviewed (approved):
Tested. Works nicely.

After merge, we should do the following:

- [x] setup quay.io builds
- [x] update container image references
- [x] add a small README or howto somewhere (basically the instructions from the PR)

timogoebel commented:
Thanks, @ohadlevy.

ShimShtein commented:
> add a small README or howto somewhere (basically the instructions from the PR)

How about adding it to [/developer_docs](https://github.com/theforeman/foreman/tree/develop/developer_docs)?

lzap commented:
Are we gonna publish also from stable branches? This will be very useful to reproduce steps with particular (older) release. Good job everyone! Nice.

ohadlevy commented:
@lzap this will require we push the dockerfile into the stable branches,
lets wait a few days before doing so, but quay is configured to try and
build images on every branch on the foreman repo on incoming git push.

On Tue, May 21, 2019 at 12:19 PM Lukáš Zapletal <notifications@github.com>
wrote:

> Are we gonna publish also from stable branches? This will be very useful
> to reproduce steps with particular (older) release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/6730?email_source=notifications&email_token=AAAE5R7AW2NPWXDK6APVFGDPWO5CLA5CNFSM4HJ4SB7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV3JOZA#issuecomment-494311268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAE5R5DWGDYSBRR6ZB7HHLPWO5CLANCNFSM4HJ4SB7A>
> .
>

ekohl commented:
I would like to have clear markers that it's experimental. Regular users should not get the impression that this is production quality just yet.

ohadlevy commented:
On Tue, May 21, 2019 at 3:25 PM Ewoud Kohl van Wijngaarden <
notifications@github.com> wrote:

> I would like to have clear markers that it's experimental. Regular users
> should not get the impression that this is production quality just yet.
>

what exactly would you like to change? we are not adding it to any user
facing documentation?

>

ekohl commented:
As discussed over IRC: https://quay.io/repository/foreman/foreman?tab=info now mentions it's experimental which sets the right expectations.

ohadlevy commented:
> > add a small README or howto somewhere (basically the instructions from the PR)
> 
> How about adding it to [/developer_docs](https://github.com/theforeman/foreman/tree/develop/developer_docs)?

I've created https://community.theforeman.org/t/dockerfile-is-now-included-in-foreman-core/13987?u=ohadlevy - I'll follow up with a update to the docs as well.

ohadlevy commented:
added documentation pr at https://github.com/theforeman/foreman/pull/6786

iNecas commented on .dockerignore:
```diff
@@ -0,0 +1,59 @@
+**/node_modules
```
@ohadlevy question: How does `.dockerignore` differ from `.gitignore`? Wouldn't we be ok with just making the `.dockerignore` a symling to `.gitignore`?

ohadlevy replied:
they are nearly identical, but with minor differences afair the main differences are the bundle/*.local and .git/.gitignore etc.

in theory, we could remove the tests folder and other things for production image.
it would be nice to keep the common parts in some shared file instead.

ohadlevy replied:
also, some differences in syntax that I've found out the hard way, see https://zzz.buzz/2018/05/23/differences-of-rules-between-gitignore-and-dockerignore/
